google-cloud-storage/lib/google/cloud/storage/hmac_key.rb

```diff
@@ -0,0 +1,312 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "google/cloud/storage/hmac_key/list"
+
+module Google
+  module Cloud
+    module Storage
+      ##
+      # # HmacKey
+      #
+      # Represents the metadata for an HMAC key, and also includes the key's
+      # secret if returned by the {Project#create_hmac_key} creation method.
+      #
+      # @see https://cloud.google.com/storage/docs/migrating#migration-simple
+      #   Migrating from Amazon S3 to Cloud Storage - Simple migration
+      #
+      # @attr_reader [String, nil] secret HMAC secret key material, or `nil` if
+      #   the key was not returned by the {Project#create_hmac_key} creation
+      #   method.
+      #
+      # @example
+      #   require "google/cloud/storage"
+      #
+      #   storage = Google::Cloud::Storage.new
+      #
+      #   service_account_email = "my_account@developer.gserviceaccount.com"
+      #   hmac_key = storage.create_hmac_key service_account_email
+      #   hmac_key.secret # ...
+      #
+      #   hmac_key = storage.hmac_key hmac_key.access_id
+      #   hmac_key.secret # nil
+      #
+      #   hmac_key.inactive!
+      #   hmac_key.delete
+      #
+      class HmacKey
+        ##
+        # @private The Connection object.
+        attr_accessor :service
+
+        ##
+        # @private The Google API Client object.
+        attr_accessor :gapi
+
+        ##
+        # @private A boolean value or a project ID string to indicate the
+        # project to be billed for operations.
+        attr_accessor :user_project
+
+        attr_reader :secret
+
+        ##
+        # @private Creates a Notification object.
+        def initialize
+          @bucket = nil
+          @service = nil
+          @gapi = nil
+          @user_project = nil
+        end
+
+        ##
+        # The ID of the HMAC Key.
+        #
+        # @return [String]
+        #
+        def access_id
+          @gapi.access_id
+        end
+
+        ##
+        # A URL that can be used to access the HMAC key using the REST API.
+        #
+        # @return [String]
+        #
+        def api_url
+          @gapi.self_link
+        end
+
+        ##
+        # Creation time of the HMAC key.
+        #
+        # @return [String]
+        #
+        def created_at
+          @gapi.time_created
+        end
+
+        ##
+        # HTTP 1.1 Entity tag for the HMAC key.
+        #
+        # @return [String]
+        #
+        def etag
+          @gapi.etag
+        end
+
+        ##
+        # The ID of the HMAC key, including the Project ID and the Access ID.
+        #
+        # @return [String]
+        #
+        def id
+          @gapi.id
+        end
+
+        ##
+        # Project ID owning the service account to which the key authenticates.
+        #
+        # @return [String]
+        #
+        def project_id
```

Comment:
This should use the `project_id` returned from the `service.get_hmac_key` instead of `@gapi.project_id` otherwise the user provided `project_id` is lost.

Reply:
The `@gapi` instance variable here holds the `Google::Apis::StorageV1::HmacKeyMetadata` instance returned by the RPCs (including `service.get_hmac_key`), so that's OK isn't it?

Reply:
oh. yes. Apologies @quartzmo, I completely misunderstood the gapi comment above. 

`@private The Google API Client object.` I thought it was a Google API client instance and not the HMACKeyMetadata instance.